src/field_theory/galois_field.lean

```diff
@@ -0,0 +1,178 @@
+  rw [derivative_sub, derivative_pow, derivative_X, h],
+  ring,
+end
+
+section splitting_field_facts
+
+variables {K : Type*} {L : Type*} [field K] [field L] {f : polynomial K}
+
+lemma exists_finset_of_splits (i : K →+* L) (sep : separable f) : splits i f →
+  ∃ (s : finset L), f.map i = C (i f.leading_coeff) *
+  (s.prod (λ a : L, (X : polynomial L) - C a)) :=
+begin
+  classical,
+  intro sp, cases exists_multiset_of_splits i sp with s h, use s.to_finset, rw h,
+  rw finset.prod_eq_multiset_prod, rw ← multiset.to_finset_eq,
+  apply nodup_of_separable_prod,
+  apply separable.of_mul_right,
+  rw ←h,
+  exact sep.map,
+end
+
+open_locale big_operators
+variables {α : Type*} (s : finset α) (g : α → polynomial K)
+
+-- I have a proof in another project, hope to PR soon - Aaron Anderson
+lemma nat_degree_prod_eq' (h : ∏ i in s, (g i).leading_coeff ≠ 0) :
+  (s.prod g).nat_degree = ∑ i in s, (g i).nat_degree :=
+begin
+  rw finset.prod_ne_zero_iff at h,
+  rw nat_degree_prod,
+  intros i hi hgi,
+  apply h i hi,
+  rw [hgi, polynomial.leading_coeff_zero],
+end
+
+lemma card_roots_eq_of_separable_of_splits
+  (sep : separable f) (sp : polynomial.splits (ring_hom.id K) f) :
+  f.roots.card = f.nat_degree :=
+begin
+  classical,
+  have hf : f ≠ 0, { rintro rfl, exact not_separable_zero sep },
+  have hflc : f.leading_coeff ≠ 0, { rwa [ne.def, leading_coeff_eq_zero] },
+  cases (exists_finset_of_splits (ring_hom.id K) sep sp) with s hs,
+  simp only [ring_hom.id_apply, map_id] at hs, rw hs,
+  rw [roots_C_mul _ hflc, roots_prod_X_sub_C, nat_degree_C_mul_eq_of_no_zero_divisors hflc],
+  rw nat_degree_prod_eq',
+  { simp only [mul_one, nat_degree_X_sub_C, nat.cast_id, finset.sum_const, nsmul_eq_mul,
+      finset.card_def], },
+  { simp only [(monic_X_sub_C _).leading_coeff, finset.prod_const_one,
+                ne.def, not_false_iff, one_ne_zero] }
+end
+
+-- linter complains, but removing causes `instance : fintype (galois_field p n)` below to fail
+instance is_splitting_field.finite_dimensional  [algebra K L] [is_splitting_field K L f] :
+  finite_dimensional K L :=
+is_splitting_field.finite_dimensional L f
+
+end splitting_field_facts
+
+/-- A finite field with `p ^ n` elements.
+Every field with the same cardinality is (non-canonically)
+isomorphic to this field. -/
+@[derive field]
+def galois_field (p : ℕ) [fact p.prime] (n : ℕ) :=
+splitting_field (X^(p^n) - X : polynomial (zmod p))
+
+instance : inhabited (@galois_field 2 (fact.mk nat.prime_two) 1) :=
+⟨37⟩
+
+namespace galois_field
+variables (p : ℕ) [fact p.prime] (n : ℕ)
+
+instance : algebra (zmod p) (galois_field p n) :=
+splitting_field.algebra _
+
+instance : is_splitting_field (zmod p) (galois_field p n) (X^(p^n) - X) :=
+polynomial.is_splitting_field.splitting_field _
+
+instance : char_p (galois_field p n) p :=
+(algebra.char_p_iff (zmod p) (galois_field p n) p).mp (by apply_instance)
+
+-- should be able to apply_instance from finite_dimensional.fintype on finite.lean
```

Comment:
if we make `galois_field` reducible this works without the above instance; is this a good idea? (or at least doing it locally)